### ISSUE TITLE
Fix threadsafe issue of ScanProperties

### ISSUE THREAD

**Issue by 304NotModified** [NLog/NLog#1262]
**NLog version**:  current (4.3 beta)

unit test: NLog.UnitTests.LogManagerTests.ThreadSafe_getCurrentClassLogger_test
```
Microsoft.Practices.Unity.ResolutionFailedException : Resolution of the dependency failed, type = "NLog.UnitTests.LogManagerTests+ClassA", name = "(none)".

Exception occurred while: Calling constructor NLog.UnitTests.LogManagerTests+ClassC(NLog.UnitTests.LogManagerTests+ClassD dd).

Exception is: TypeInitializationException - The type initializer for 'ClassC' threw an exception.

-----------------------------------------------

At the time of the exception, the container was:



  Resolving NLog.UnitTests.LogManagerTests+ClassA,(none)

  Resolving parameter "dd" of constructor NLog.UnitTests.LogManagerTests+ClassA(NLog.UnitTests.LogManagerTests+ClassB dd)

    Resolving NLog.UnitTests.LogManagerTests+ClassB,(none)

    Resolving parameter "dd" of constructor NLog.UnitTests.LogManagerTests+ClassB(NLog.UnitTests.LogManagerTests+ClassC dd)

      Resolving NLog.UnitTests.LogManagerTests+ClassC,(none)

      Calling constructor NLog.UnitTests.LogManagerTests+ClassC(NLog.UnitTests.LogManagerTests+ClassD dd)



---- System.TypeInitializationException : The type initializer for 'ClassC' threw an exception.

-------- System.InvalidOperationException : Collection was modified after the enumerator was instantiated.
```

See https://ci.appveyor.com/project/nlog/nlog/build/4.0.1976/tests

---

   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)

   at System.Collections.Generic.Queue`1.Enumerator.MoveNext()

   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)

   at NLog.Internal.ObjectGraphScanner.ScanProperties[T](List`1 result, Object o, Int32 level, HashSet`1 visitedObjects) in C:\projects\nlog\src\NLog\Internal\ObjectGraphScanner.cs:line 129

   at NLog.Internal.ObjectGraphScanner.ScanProperties[T](List`1 result, Object o, Int32 level, HashSet`1 visitedObjects) in C:\projects\nlog\src\NLog\Internal\ObjectGraphScanner.cs:line 152

   at NLog.Internal.ObjectGraphScanner.FindReachableObjects[T](Object[] rootObjects) in C:\projects\nlog\src\NLog\Internal\ObjectGraphScanner.cs:line 67

   at NLog.Internal.TargetWithFilterChain.PrecalculateStackTraceUsage() in C:\projects\nlog\src\NLog\Internal\TargetWithFilterChain.cs:line 95

   at NLog.LogFactory.GetTargetsByLevelForLogger(String name, IEnumerable`1 rules, TargetWithFilterChain[] targetsByLevel, TargetWithFilterChain[] lastTargetsByLevel, Boolean[] suppressedLevels) in C:\projects\nlog\src\NLog\LogFactory.cs:line 769

   at NLog.LogFactory.GetTargetsByLevelForLogger(String name, IEnumerable`1 rules, TargetWithFilterChain[] targetsByLevel, TargetWithFilterChain[] lastTargetsByLevel, Boolean[] suppressedLevels) in C:\projects\nlog\src\NLog\LogFactory.cs:line 760

   at NLog.LogFactory.GetConfigurationForLogger(String name, LoggingConfiguration configuration) in C:\projects\nlog\src\NLog\LogFactory.cs:line 782

   at NLog.LogFactory.GetLogger(LoggerCacheKey cacheKey) in C:\projects\nlog\src\NLog\LogFactory.cs:line 951

   at NLog.LogFactory.GetLogger(String name) in C:\projects\nlog\src\NLog\LogFactory.cs:line 409

   at NLog.LogManager.GetCurrentClassLogger() in C:\projects\nlog\src\NLog\LogManager.cs:line 175

   at NLog.UnitTests.LogManagerTests.ClassC..cctor() in C:\projects\nlog\tests\NLog.UnitTests\LogManagerTests.cs:line 577

**Comment by 304NotModified** [comment]
https://ci.appveyor.com/project/nlog/nlog/build/4.0.1982

**Comment by 304NotModified** [comment]
https://ci.appveyor.com/project/nlog/nlog/build/4.0.2049

**Comment by 304NotModified** [comment]
Still a problem:

https://ci.appveyor.com/project/nlog/nlog/build/4.0.2106 (up-to-date with master)

**Comment by 304NotModified** [comment]
https://ci.appveyor.com/project/nlog/nlog/build/4.0.2148

**Comment by 304NotModified** [comment]
Fixed in #1310
